### PR TITLE
[NPM] Caching ability for Pods and NS

### PR DESCRIPTION
**Pull request by vakalapa**
2nd try: This PR includes below changes:

- Adding Caching ability for updated Pods to help relieve pressure on update events
- optimizing named ports code path for uniformity
- Removing generic PodMap and maintaining a NameSpace specific podmap
- ResourceVersions checks for Namespaces, Pods and Network Policies

### CODE REVIEW

File: npm/pod.go [hunk 1]
```diff
@@ -32,222 +94,371 @@
...
 		return nil
 	}
 
+	// Ignore adding the HostNetwork pod to any ipsets.
+	if isHostNetworkPod(podObj) {
+		log.Logf("HostNetwork POD IGNORED: [%s/%s/%s/%+v%s]", podObj.GetObjectMeta().GetUID(), podObj.Namespace, podObj.Name, podObj.Labels, podObj.Status.PodIP)
+		return nil
+	}
+
+	if podObj.Status.Phase == v1.PodSucceeded || podObj.Status.Phase == v1.PodFailed {
```
Comment (neaggarwMS) [inline npm/pod.go:168]: Not sure if I understood this change, why are we returning on PodSucceeded?

File: npm/pod.go [hunk 2]
```diff
@@ -32,222 +94,374 @@
...
```
Comment (neaggarwMS) [inline npm/pod.go:168]: Can you add a comment for why this is blocked?

Comment (vakalapa) [reply npm/pod.go:170]: Done

File: npm/pod.go (hunk 1 above)
Comment (vakalapa) [reply npm/pod.go:170]: As we discussed offline, k8s considers Succeeded or Failed pod as terminated and will not restart them.

**Review by neaggarwMS — APPROVED** [review]
:shipit:

File: npm/pod.go [hunk 3]
```diff
@@ -32,222 +94,378 @@
...
+		}
+		// Add the pod to its namespace's ipset.
+		if err = ipsMgr.AddToSet(newPodObjNs, newPodObjIP, util.IpsetNetHashFlag, cachedPodObj.podUID); err != nil {
+			log.Errorf("Error: failed to add pod to namespace ipset.")
+		}
+	} else {
+		//if no change in labels then return
+		if reflect.DeepEqual(cachedLabels, newPodObjLabel) {
+			log.Logf(
+				"POD UPDATING:\n nothing to delete or add. pod: [%s/%s]",
```
Comment (matmerr) [inline npm/pod.go:356]: newline?

**Pull request merged**